File: android/src/main/java/dev/sankofa/rn/SankofaDeployBundleProvider.kt
package dev.sankofa.rn

import android.content.Context
import java.io.File

/**
 * Bundle provider for Expo prebuild/dev-client and bare React Native hosts.
 *
 * In release builds, override ReactNativeHost.getJSBundleFile() and return
 * this value when non-null:
 *
 *   override fun getJSBundleFile(): String? =
 *     SankofaDeployBundleProvider.getJSBundleFile(applicationContext)
 *       ?: super.getJSBundleFile()
 */
object SankofaDeployBundleProvider {
  /** SharedPreferences flag flipped to `true` on the first invocation —
   *  proves the host's MainApplication actually wired its
   *  `getJSBundleFile` override into us. `checkIntegration()` reads
   *  this to tell the host "your bundle loader patch is/isn't live."
   */
  private const val WIRED_FLAG = "bundle_loader_wired"

  @JvmStatic
  fun getJSBundleFile(context: Context): String? {
    context
      .getSharedPreferences("sankofa_deploy", Context.MODE_PRIVATE)
      .edit()
      .putBoolean(WIRED_FLAG, true)
      .apply()
    val path = context
      .getSharedPreferences("sankofa_deploy", Context.MODE_PRIVATE)
      .getString("bundle_path", null)
      ?: return null
    return if (File(path).exists()) path else null
  }
}

File: ios/SankofaDeployBundleProvider.swift
import Foundation

@objc(SankofaDeployBundleProvider)
public final class SankofaDeployBundleProvider: NSObject {
  /// UserDefaults flag flipped to `true` on the first invocation —
  /// proves the host's AppDelegate / ReactNativeDelegate actually
  /// wired its `bundleURL()` override into us. `checkIntegration()`
  /// reads this to tell the host "your bundle loader patch is/isn't live."
  private static let wiredFlagKey = "sankofa_deploy_bundle_loader_wired"

  @objc
  public static func bundleURL() -> URL? {
    UserDefaults.standard.set(true, forKey: wiredFlagKey)
    guard let path = UserDefaults.standard.string(forKey: "sankofa_deploy_bundle_path"),
          FileManager.default.fileExists(atPath: path) else {
      return nil
    }
    return URL(fileURLWithPath: path)
  }
}
